src/lib/y2partitioner/actions/resize_partition.rb

```diff
@@ -0,0 +1,135 @@
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "yast/i18n"
+require "y2partitioner/dialogs/partition_resize"
+require "y2partitioner/ui_state"
+
+Yast.import "Popup"
+
+module Y2Partitioner
+  module Actions
+    # Action for resizing a partition
+    class ResizePartition
+      include Yast::I18n
+
+      # Constructor
+      #
+      # @param partition [Y2Storage::Partition]
+      def initialize(partition)
+        textdomain "storage"
+
+        @partition = partition
+        @resize_info = partition.detect_resize_info
+        UIState.instance.select_row(partition)
+      end
+
+      # Checks whether it is possible to resize the partition, and if so,
+      # the action is performed.
+      #
+      # @note An error popup is shown when the partition cannot be resized.
+      #
+      # @return [Symbol, nil]
+      def run
+        return :back unless validate
+        resize
+      end
+
+    private
+
+      # @return [Y2Storage::Partition] partition to resize
+      attr_reader :partition
+
+      # @return [Y2Storage::ResizeInfo]
+      attr_reader :resize_info
+
+      # Runs the dialog to resize the partition
+      #
+      # @return [Symbol] :finish if the dialog returns :next; dialog result otherwise.
+      def resize
+        result = Dialogs::PartitionResize.run(partition, resize_info)
+        fix_end_alignment(resize_info)
+
+        result == :next ? :finish : result
+      end
+
+      # Checks whether the resize action can be performed
+      #
+      # @see Y2Storage::ResizeInfo#resize_ok?
+      #
+      # @return [Boolean] true if the resize action can be performed; false otherwise.
+      def validate
+        return true if resize_info.resize_ok?
+
+        # TODO: Distinguish the reason why it is not possible to resize, for example:
+        # * partition used by commited LVM or MD RAID
+        # * extended partition with committed logical partitions
+
+        Yast::Popup.Error(
+          # TRANSLATORS: an error popup message
+          _("This partition cannot be resized.")
+        )
+
+        false
+      end
+
+      # After the partition's size was changed during resizing, make sure the
+      # new size meets all alignment requirements, but is still between
+      # min_size and max_size. This may change the partition's size (and
+      # region).
+      #
+      # @param resize_info [ResizeInfo]
+      def fix_end_alignment(resize_info)
+        return if partition.nil?
+
+        ptable = partition.partition_table
+        return if !ptable.require_end_alignment? || partition.end_aligned?
```

Comment:
So we are aligning only if absolutely needed (DASD). Don't we want to align basically always than is possible?

Reply:
I'm asking because the partitioner already forces alignments of the partition starts. So if a user creates N partitions in a row using a big space and we align all the starts but none of the endings we are introducing gaps that make very little sense.